src/toast/ops/sim_cosmic_rays.py

```diff
@@ -0,0 +1,294 @@
+    )
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def load_cosmic_ray_data(self, filename):
+        data_dic = np.load(filename)
+
+        return data_dic
+
+    def resample_cosmic_ray_statistics(self, arr, Nresamples, key, counter):
+
+        resampled = np.zeros((Nresamples, arr.shape[1]))
+
+        for ii in range(arr.shape[1]):
+            # Resample by considering the bulk of  the Parameter distribution ~2sigma central interval
+            bins = np.linspace(
+                np.quantile(
+                    arr[:, ii],
+                    0.025,
+                ),
+                np.quantile(arr[:, ii], 0.975),
+                30,
+            )
+
+            binned, edges = np.histogram(arr[:, ii], bins=bins)
+
+            xb = 0.5 * (edges[:-1] + edges[1:])
+            CDF = np.cumsum(binned) / binned.sum()
+
+            pinv = interpolate.interp1d(CDF, xb, fill_value="extrapolate")
+            # r = np.random.rand(Nresamples)
+            r = rng.random(
+                Nresamples,
+                sampler="uniform_01",
+                key=key,
+                counter=counter,
+            )
+
+            resampled[:, ii] = pinv(r)
+
+        return resampled
+
+    @function_timer
+    def _exec(self, data, detectors=None, **kwargs):
+        env = Environment.get()
+        log = Logger.get()
+        if self.crfile is None:
+            raise AttributeError(
+                "OpInjectCosmicRays cannot run if you don't provide cosmic ray data."
+            )
+        for ob in data.obs:
+            # Get the detectors we are using for this observation
+            dets = ob.select_local_detectors(detectors)
+            if len(dets) == 0:
+                # Nothing to do for this observation
+                continue
+            comm = ob.comm
+            rank = ob.comm_rank
+            # Make sure detector data output exists
+            ob.detdata.ensure(self.det_data, detectors=dets)
+            obsindx = ob.uid
+            telescope = ob.telescope.uid
+            focalplane = ob.telescope.focalplane
+            size = ob.detdata[self.det_data][dets[0]].size
+            samplerate = focalplane.sample_rate.to_value(u.Hz)
+
+            obstime_seconds = size / samplerate
+            n_events_expected = self.eventrate * obstime_seconds
+            key1 = self.realization * 4294967296 + telescope * 65536
+            counter2 = 0
+
+            for kk, det in enumerate(dets):
+                detindx = focalplane[det]["uid"]
+                key2 = obsindx
+                counter1 = detindx
+
+                rngdata = rng.random(
+                    size,
+                    sampler="gaussian",
+                    key=(key1, key2),
+                    counter=(counter1, counter2),
+                )
+                filename = self.crfile.replace("detector", f"det{kk}")
+                data_dic = self.load_cosmic_ray_data(filename)
```

Comment:
This is fine for now, but imagine we had a job running with 1000 detectors for a year, with one observation per hour.  This job would load per-detector files 8760000 times from across all processes.  We should eventually look at how big these files are and see if we should put all the information into one file and load it once and broadcast it to all detectors.

Reply:
Totally agree with this.  the way the CRs are currently  injected is by inspecting the  detector properties stored on each data file.  however, the glitch statistical properties  as well as the wafer noise and the correlation matrix can be easily extended to all the detectors by simply reading them from a single file and broadcast it. On the other hand, we want to leave the user the possibility to inject CRs as they prefer. Can include an extra-trait taking care of  this.. what do you think?